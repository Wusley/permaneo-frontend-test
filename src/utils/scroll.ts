export default class Scroll {
  static scrollToTop() {
    window.scrollTo( { top: 0, behavior: 'smooth' } )
  }
}
