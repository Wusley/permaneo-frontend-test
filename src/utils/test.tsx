import { configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import { renderHook } from '@testing-library/react'
import '@testing-library/jest-dom'
import Providers from '@/providers'
import { rootReducer, RootState } from '@/redux/store'

const createTestStore = ( preloadedState: Partial<RootState> = {} ) =>
  configureStore( {
    reducer: rootReducer,
    preloadedState,
    middleware: ( getDefaultMiddleware ) => getDefaultMiddleware()
  } )

const renderWithStore = (
  ui: React.ReactElement,
  preloadedState: Partial<RootState> = {},
  options?: unknown
) => {
  const store = createTestStore( preloadedState )

  return {
    ...render( ui, {
      wrapper: ( { children } ) => <Providers store={store}>{children}</Providers>,
      ...options
    } ),
    store
  }
}

const renderHookWithStore = <TProps, TResult>(
  hook: ( props: TProps ) => TResult,
  preloadedState: Partial<RootState> = {}
) => {
  const store = createTestStore( preloadedState )

  return {
    ...renderHook( hook, {
      wrapper: ( { children } ) => <Providers store={store}>{children}</Providers>
    } ),
    store
  }
}

export * from '@testing-library/react'
export { renderWithStore, renderHookWithStore }
