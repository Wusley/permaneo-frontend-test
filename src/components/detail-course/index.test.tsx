import { screen, renderWithStore } from '@/utils/test'
import { DetailCourse } from './'
import moment from 'moment'

const mockFormatCurrency = jest.fn( ( price ) => price.toFixed( 2 ) )
jest.mock( '@/hooks/useNumberFormatter', () => () => mockFormatCurrency )

jest.mock( 'moment', () => {
  const actualMoment = jest.requireActual( 'moment' )
  actualMoment.defineLocale = jest.fn()
  return ( dateString: string ) => actualMoment.utc( dateString )
} )

describe( 'DetailCourse', () => {
  const defaultProps = {
    id: 1,
    title: 'Curso de TypeScript',
    description: 'Aprenda TypeScript do básico ao avançado.',
    startAt: '2025-01-20',
    price: 199.99
  }

  it( 'renders the course details correctly', () => {
    const container = renderWithStore( <DetailCourse {...defaultProps} /> )

    expect( screen.getByRole( 'heading', { level: 2 } ) ).toHaveTextContent( defaultProps.title )
    expect( screen.getByText( defaultProps.description ) ).toBeInTheDocument()
    expect( screen.getByText( `R$ ${mockFormatCurrency( defaultProps.price )}` ) ).toBeInTheDocument()
    expect( screen.getByText( `Esse curso começa em ${moment( defaultProps.startAt ).format( 'DD/MM/YYYY' )}` ) ).toBeInTheDocument()
    expect( container ).toMatchSnapshot()
  } )

  it( 'renders the favorite button', () => {
    renderWithStore( <DetailCourse {...defaultProps} /> )
    expect( screen.getByRole( 'button' ) ).toBeInTheDocument()
  } )

  it( 'renders the subscription button', () => {
    renderWithStore( <DetailCourse {...defaultProps} /> )
    expect( screen.getByText( /inscreva-se agora/i ) ).toBeInTheDocument()
  } )
} )
