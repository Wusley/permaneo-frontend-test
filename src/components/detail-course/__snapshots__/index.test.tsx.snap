// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailCourse renders the course details correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div>
      <section
        class="css-lcmtu0"
      >
        <div
          class="css-69i1ev"
        >
          <h2
            class="chakra-heading css-1xjursl"
          >
            Curso de TypeScript
          </h2>
          <button
            aria-label="Favoritar curso"
            class="chakra-button css-zzezgs"
            data-loading="false"
            type="button"
          >
            <svg
              aria-hidden="true"
              color="#F56565"
              fill="currentColor"
              focusable="false"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(245, 101, 101);"
              viewBox="0 0 24 24"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
          </button>
        </div>
        <p
          class="chakra-text css-1wm9mbs"
        >
          Aprenda TypeScript do básico ao avançado.
        </p>
        <div
          class="chakra-stack css-1ch3xf2"
        >
          <div
            class="css-j7qwjs"
          >
            <strong
              class="chakra-text css-9hpz1u"
            >
              R$ 
              199.99
            </strong>
            <p
              class="chakra-text css-1u3k8gl"
            >
              Esse curso começa em 
              20/01/2025
            </p>
          </div>
          Inscreva-se agora
        </div>
      </section>
      <span
        hidden=""
        id="__chakra_env"
      />
    </div>
    <div
      class="chakra-portal"
    >
      <div
        aria-label="Notifications-top"
        aria-live="polite"
        id="chakra-toast-manager-top"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-top-left"
        aria-live="polite"
        id="chakra-toast-manager-top-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-top-right"
        aria-live="polite"
        id="chakra-toast-manager-top-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom-left"
        aria-live="polite"
        id="chakra-toast-manager-bottom-left"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
      <div
        aria-label="Notifications-bottom"
        aria-live="polite"
        id="chakra-toast-manager-bottom"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <div
        aria-label="Notifications-bottom-right"
        aria-live="polite"
        id="chakra-toast-manager-bottom-right"
        role="region"
        style="position: fixed; z-index: var(--toast-z-index, 5500); pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
  </body>,
  "container": <div>
    <section
      class="css-lcmtu0"
    >
      <div
        class="css-69i1ev"
      >
        <h2
          class="chakra-heading css-1xjursl"
        >
          Curso de TypeScript
        </h2>
        <button
          aria-label="Favoritar curso"
          class="chakra-button css-zzezgs"
          data-loading="false"
          type="button"
        >
          <svg
            aria-hidden="true"
            color="#F56565"
            fill="currentColor"
            focusable="false"
            height="28"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(245, 101, 101);"
            viewBox="0 0 24 24"
            width="28"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
            />
          </svg>
        </button>
      </div>
      <p
        class="chakra-text css-1wm9mbs"
      >
        Aprenda TypeScript do básico ao avançado.
      </p>
      <div
        class="chakra-stack css-1ch3xf2"
      >
        <div
          class="css-j7qwjs"
        >
          <strong
            class="chakra-text css-9hpz1u"
          >
            R$ 
            199.99
          </strong>
          <p
            class="chakra-text css-1u3k8gl"
          >
            Esse curso começa em 
            20/01/2025
          </p>
        </div>
        Inscreva-se agora
      </div>
    </section>
    <span
      hidden=""
      id="__chakra_env"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
